sklearn/mixture/gmm.py

```diff
@@ -112,7 +114,157 @@ def sample_gaussian(mean, covar, covariance_type='diag', n_samples=1,
     return (rand.T + mean).T
 
 
-class GMM(BaseEstimator):
+def _check_X(X, n_components=None, n_features=None):
+    """Check the input data X.
+
+    Parameters
+    ----------
+    X : array-like, (n_samples, n_features)
+
+    n_components : int
+
+    Returns
+    -------
+    X : array, (n_samples, n_features)
+    """
+    X = check_array(X, dtype=[np.float64, np.float32], ensure_2d=False)
+    if X.ndim != 2:
+        raise ValueError("Expected the input data X have 2 dimensions, "
+                         "but got %s dimension(s)" %
+                         X.ndim)
+    if n_components is not None and X.shape[0] < n_components:
+        raise ValueError('Expected n_samples >= n_components'
+                         'but got n_components = %d, n_samples = %d'
+                         % (n_components, X.shape[0]))
+    if n_features is not None and X.shape[1] != n_features:
+        raise ValueError("Expected the input data X have %d features, "
+                         "but got %d features"
+                         % (n_features, X.shape[1]))
+    return X
+
+
+class _MixtureBase(six.with_metaclass(ABCMeta, DensityMixin, BaseEstimator)):
+    """Base class for mixture models.
+
+    This abstract class specifies the interface by abstract methods and
+    provides basic common methods for mixture models.
+    """
+    def __init__(self, n_components, covariance_type, tol, reg_covar,
+                 max_iter, n_init, init_params, random_state, verbose):
+        self.n_components = n_components
+        self.covariance_type = covariance_type
+        self.tol = tol
+        self.reg_covar = reg_covar
+        self.random_state = random_state
+        self.max_iter = max_iter
+        self.n_init = n_init
+        self.init_params = init_params
+        self.verbose = verbose
+
+        self.n_features_ = None
+        self.converged_ = False
+
+        if n_init < 1:
```

Comment:
We usually avoid checking the parameters in `__init__`, since a parameter can also be changed with `set_params`